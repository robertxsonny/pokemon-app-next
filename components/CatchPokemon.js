import { useEffect, useState } from 'react';
import styled from '@emotion/styled';
import Link from 'next/link';

import Modal from './Modal';
import { SentenceCase, SmallCtaButton, TextButton } from '../styles/shared';

import { useMyPokemons } from '../hooks/my-pokemon';

const SecondaryTextButton = styled(TextButton)({
  color: 'green',
  marginTop: 12
})

const SecondaryLinkButton = SecondaryTextButton.withComponent('a')

const DangerTextButton = styled(TextButton)({
  color: 'red',
  marginTop: 12
})

const DangerCtaButton = styled(SmallCtaButton)({
  backgroundColor: 'red',
  '&:hover:not([disabled])': {
    backgroundColor: 'firebrick'
  },
  '&:active:not([disabled])': {
    backgroundColor: 'darkred'
  }
})

const TextInput = styled.input({
  marginBottom: 16,
  width: '100%',
  boxSizing: 'border-box',
  padding: '6px 0',
  border: 'none',
  outline: 'none',
  borderBottom: '2px solid dimgray',
  transition: '0.3s all ease',
  textTransform: 'capitalize',
  '&:focus': {
    borderBottom: '2px solid limegreen',
  }
})

const CatchPokemon = ({ open, name, onCaught, onClose }) => {
  const { hasPokemonWithNickname, catchPokemon } = useMyPokemons();
  const [nickname, setNickname] = useState();
  const [canCatch, setCanCatch] = useState(null);
  const [caught, setCaught] = useState(false);

  useEffect(() => {
    const initNickname = () => {
      if (!hasPokemonWithNickname(name)) {
        setNickname(name);
      } else {
        let lastNum = 1;
        while (hasPokemonWithNickname(`${name}${lastNum}`)) {
          lastNum = lastNum + 1;
        }
        setNickname(`${name}${lastNum}`)
      }
    }
  
    if (open && canCatch === null) {
      setCaught(false);
      setTimeout(() => {
        const success = Math.random() < 0.5;
        if (success) {
          initNickname();
        }
        setCanCatch(success);
      }, 3000)
    } else if (!open) {
      setCanCatch(null);
    }
  }, [open, canCatch, hasPokemonWithNickname])

  const onNicknameChange = (event) => {
    const { value } = event.target;
    setNickname(value.toLowerCase());
  }

  const onCatchPokemon = () => {
    catchPokemon(name, nickname);
    setCaught(true);
    if (onCaught) {
      onCaught();
    }
  }

  const nameWihSentenceCase = <SentenceCase>{name}</SentenceCase>

  if (caught) {
    return (
      <Modal open={open} title={<><SentenceCase>{nickname}</SentenceCase> has been added to your collection</>} emoji="✅">
        <Link href="/my-pokemon" passHref>
          <SecondaryLinkButton>View my collection</SecondaryLinkButton>
        </Link>
        <SecondaryTextButton onClick={onClose}>Go back</SecondaryTextButton>
      </Modal>
    )
  }

  if (canCatch === true) {
    return (
      <Modal open={open} title={<>{nameWihSentenceCase} was caught!</>} emoji="👍">
        <p>Give it a unique nickname to save it to your collection</p>
        <TextInput value={nickname} onChange={onNicknameChange} />
        <SmallCtaButton disabled={!nickname || hasPokemonWithNickname(nickname)} onClick={onCatchPokemon}>Save to my collection</SmallCtaButton>
        <SecondaryTextButton onClick={onClose}>Cancel</SecondaryTextButton>
      </Modal>
    )
  }

  if (canCatch === false) {
    return (
      <Modal open={open} title={<>Oops, failed to catch {nameWihSentenceCase}!</>} emoji="😞">
        <DangerCtaButton onClick={() => setCanCatch(null)}>Try again</DangerCtaButton>
        <DangerTextButton onClick={onClose}>Go back</DangerTextButton>
      </Modal>
    )
  }

  return (
    <Modal open={open} title={<>Trying to catch {nameWihSentenceCase}</>} emoji="⏳">
      <p>Please wait...</p>
    </Modal>
  )
}

export default CatchPokemon;
