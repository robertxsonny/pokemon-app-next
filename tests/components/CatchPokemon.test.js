import { render, screen, fireEvent, act } from '@testing-library/react';
import CatchPokemon from '../../components/CatchPokemon';

jest.mock('../../hooks/my-pokemon', () => ({
  useMyPokemons: () => ({
    hasPokemonWithNickname: jest.fn(name => name === 'squirtle'),
    catchPokemon: jest.fn()
  })
}))

describe('CatchPokemon loading state', () => {
  test('Should initially render loading screen', () => {
    const { container } = render(<CatchPokemon open name="pikachu"/>);
    expect(container).toMatchSnapshot();
  })

  test('Should not render anything if not opened', () => {
    const { container } = render(<CatchPokemon open={false}/>);
    expect(container).toMatchSnapshot();
  })
});

describe('CatchPokemon success state', () => {
  const actualRandom = global.Math.random;

  beforeEach(() => {
    jest.useFakeTimers();
    global.Math.random = jest.fn(() => 0.1);
  })

  afterEach(() => {
    jest.useRealTimers();
    global.Math.random = actualRandom;
  })

  test('Should render saving form once it allows to catch', () => {
    const { container } = render(<CatchPokemon open name="pikachu" />);
    act(() => jest.runAllTimers());
    expect(container).toMatchSnapshot();
    expect(screen.getByDisplayValue('pikachu')).toBeTruthy();

    fireEvent.click(screen.getByRole('button', { name: 'Save to my collection' }));
    expect(container).toMatchSnapshot();
  })

  test('Should give prefilled unique nickname', () => {
    const { container } = render(<CatchPokemon open name="squirtle" />);
    act(() => jest.runAllTimers());
    expect(container).toMatchSnapshot();
    expect(screen.getByDisplayValue('squirtle1')).toBeTruthy();
  })
})

describe('CatchPokemon failed state', () => {
  const actualRandom = global.Math.random;

  beforeEach(() => {
    jest.useFakeTimers();
    global.Math.random = jest.fn(() => 0.8);
  })

  afterEach(() => {
    jest.useRealTimers();
    global.Math.random = actualRandom;
  })

  test('Should show error dialog if catching Pokemon is failed', () => {
    const { container } = render(<CatchPokemon open name="pikachu" />);
    act(() => jest.runAllTimers());
    expect(container).toMatchSnapshot();

    fireEvent.click(screen.getByRole('button', { name: 'Try again' }));
    expect(container).toMatchSnapshot();
  })
})
