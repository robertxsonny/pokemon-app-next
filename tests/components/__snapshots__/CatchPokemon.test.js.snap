// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CatchPokemon failed state Should show error dialog if catching Pokemon is failed 1`] = `
<div>
  <div
    class="css-1f6dfq1-ModalContainer e2hvhql2"
  >
    <div
      class="css-pujlrx-ModalContent e2hvhql1"
    >
      <span
        aria-label="[object Object]"
        class="css-1k4msm7-Emoji e55skjj7"
        role="img"
      >
        üòû
      </span>
      <h3
        class="css-1f0is5a-ModalTitle e2hvhql0"
      >
        Oops, failed to catch 
        <span
          class="css-8aj3s4-SentenceCase e55skjj8"
        >
          pikachu
        </span>
        !
      </h3>
      <button
        class="css-11qewao-CtaButton-SmallCtaButton-DangerCtaButton ejgy2rx1"
      >
        Try again
      </button>
      <button
        class="css-hqh9ye-TextButton-DangerTextButton ejgy2rx2"
      >
        Go back
      </button>
    </div>
  </div>
</div>
`;

exports[`CatchPokemon failed state Should show error dialog if catching Pokemon is failed 2`] = `
<div>
  <div
    class="css-1f6dfq1-ModalContainer e2hvhql2"
  >
    <div
      class="css-pujlrx-ModalContent e2hvhql1"
    >
      <span
        aria-label="[object Object]"
        class="css-1k4msm7-Emoji e55skjj7"
        role="img"
      >
        ‚è≥
      </span>
      <h3
        class="css-1f0is5a-ModalTitle e2hvhql0"
      >
        Trying to catch 
        <span
          class="css-8aj3s4-SentenceCase e55skjj8"
        >
          pikachu
        </span>
      </h3>
      <p>
        Please wait...
      </p>
    </div>
  </div>
</div>
`;

exports[`CatchPokemon loading state Should initially render loading screen 1`] = `
<div>
  <div
    class="css-1f6dfq1-ModalContainer e2hvhql2"
  >
    <div
      class="css-pujlrx-ModalContent e2hvhql1"
    >
      <span
        aria-label="[object Object]"
        class="css-1k4msm7-Emoji e55skjj7"
        role="img"
      >
        ‚è≥
      </span>
      <h3
        class="css-1f0is5a-ModalTitle e2hvhql0"
      >
        Trying to catch 
        <span
          class="css-8aj3s4-SentenceCase e55skjj8"
        >
          pikachu
        </span>
      </h3>
      <p>
        Please wait...
      </p>
    </div>
  </div>
</div>
`;

exports[`CatchPokemon loading state Should not render anything if not opened 1`] = `<div />`;

exports[`CatchPokemon success state Should give prefilled unique nickname 1`] = `
<div>
  <div
    class="css-1f6dfq1-ModalContainer e2hvhql2"
  >
    <div
      class="css-pujlrx-ModalContent e2hvhql1"
    >
      <span
        aria-label="[object Object]"
        class="css-1k4msm7-Emoji e55skjj7"
        role="img"
      >
        üëç
      </span>
      <h3
        class="css-1f0is5a-ModalTitle e2hvhql0"
      >
        <span
          class="css-8aj3s4-SentenceCase e55skjj8"
        >
          squirtle
        </span>
         was caught!
      </h3>
      <p>
        Give it a unique nickname to save it to your collection
      </p>
      <input
        class="css-9ab6cq-TextInput ejgy2rx0"
        value="squirtle1"
      />
      <button
        class="css-v2dbck-CtaButton-SmallCtaButton e55skjj0"
      >
        Save to my collection
      </button>
      <button
        class="css-qqk8k3-TextButton-SecondaryTextButton ejgy2rx3"
      >
        Cancel
      </button>
    </div>
  </div>
</div>
`;

exports[`CatchPokemon success state Should render saving form once it allows to catch 1`] = `
<div>
  <div
    class="css-1f6dfq1-ModalContainer e2hvhql2"
  >
    <div
      class="css-pujlrx-ModalContent e2hvhql1"
    >
      <span
        aria-label="[object Object]"
        class="css-1k4msm7-Emoji e55skjj7"
        role="img"
      >
        üëç
      </span>
      <h3
        class="css-1f0is5a-ModalTitle e2hvhql0"
      >
        <span
          class="css-8aj3s4-SentenceCase e55skjj8"
        >
          pikachu
        </span>
         was caught!
      </h3>
      <p>
        Give it a unique nickname to save it to your collection
      </p>
      <input
        class="css-9ab6cq-TextInput ejgy2rx0"
        value="pikachu"
      />
      <button
        class="css-v2dbck-CtaButton-SmallCtaButton e55skjj0"
      >
        Save to my collection
      </button>
      <button
        class="css-qqk8k3-TextButton-SecondaryTextButton ejgy2rx3"
      >
        Cancel
      </button>
    </div>
  </div>
</div>
`;

exports[`CatchPokemon success state Should render saving form once it allows to catch 2`] = `
<div>
  <div
    class="css-1f6dfq1-ModalContainer e2hvhql2"
  >
    <div
      class="css-pujlrx-ModalContent e2hvhql1"
    >
      <span
        aria-label="[object Object]"
        class="css-1k4msm7-Emoji e55skjj7"
        role="img"
      >
        ‚úÖ
      </span>
      <h3
        class="css-1f0is5a-ModalTitle e2hvhql0"
      >
        <span
          class="css-8aj3s4-SentenceCase e55skjj8"
        >
          pikachu
        </span>
         has been added to your collection
      </h3>
      <a
        class="css-crp5p1-SecondaryLinkButton-TextButton-SecondaryTextButton ejgy2rx4"
        href="/my-pokemon"
      >
        View my collection
      </a>
      <button
        class="css-qqk8k3-TextButton-SecondaryTextButton ejgy2rx3"
      >
        Go back
      </button>
    </div>
  </div>
</div>
`;
